concert/devices/cameras/pco.py

```diff
@@ -1,112 +1,47 @@
-        self._lock_access = True
+    async def start_readout(self):
+        self._timestamp_enabled = await self.get_timestamp() in ['both', 'binary']
+        await super().start_readout()
 
     @background
-    async def stop_recording(self):
-        self._lock_access = False
-        super().stop_recording()
-
-    async def _grab_real(self, index=None):
-        # For the PCO.4000 there must be a delay of at least 1.2 second
-        # between consecutive grabs, otherwise it crashes. We provide
-        # appropriate timeout here.
-        current = time.time()
-        if self._last_grab_time and current - self._last_grab_time < 1.2:
-            await asyncio.sleep(1.2 - current + self._last_grab_time)
-        result = await super()._grab_real(index=index)
-        self._last_grab_time = time.time()
-
-        return result
-
-    def __getattribute__(self, name):
-        if object.__getattribute__(self, '_lock_access') and name \
-           not in PCO4000._ALLOWED_DURING_RECORDING:
-            raise AttributeError("{} cannot be accessed during recording".format(name))
-        return object.__getattribute__(self, name)
-
-
-class Dimax(Pco, base.BufferedMixin):
-
-    """A pco.dimax camera implementation."""
-
-    @background
-    async def start_recording(self):
-        await super().start_recording()
-        # By low frame rates the camera returns status that it is already recording, whereas it is
-        # not. Waiting 1 frame time seems to help.
-        time.sleep((1 / self.frame_rate).to(q.s).magnitude)
-
-    async def _readout_real(self, num_frames=None):
-        """Readout *num_frames* frames."""
-        recorded_frames = await self.get_recorded_frames()
-
-        if num_frames is None:
-            num_frames = recorded_frames.magnitude
-
-        if not 0 < num_frames <= recorded_frames:
-            raise base.CameraError("Number of frames {} ".format(num_frames)
-                                   + "must be more than zero and less than the recorded "
-                                   + "number of frames {}".format(recorded_frames))
-
-        try:
-            self.uca.start_readout()
-
-            for i in range(num_frames):
-                yield await self.grab()
-        finally:
-            self.uca.stop_readout()
+    async def grab(self) -> ImageWithMetadata:
+        """Return a concert.storage.ImageWithMetadata (subclass of np.ndarray) with data of the
+        current frame."""
+        img = await self._grab_real()
+        if self._timestamp_enabled:
+            try:
+                timestamp = Timestamp(img)
+                img = self.convert(img)
+                img = img.view(ImageWithMetadata)
+                img.metadata['frame_number'] = timestamp.number
+                img.metadata['timestamp'] = timestamp.time.isoformat()
+                return img
+            except TimestampError:
+                LOG.error("Can not extract timestamp from frame.")
```

Comment:
You surely need to re-raise. Otherwise people will be very surprised what's happening. And also tests should cover the re-raise.

Reply:
But re-raising it would mean, that every frame with a broken timestamp would be lost. Do we want this?

Reply:
It would mean the addon stops the whole data acquisition, which is what you might want if time stamp is super-important.

Reply:
I explicitly set the timestamp to be there, but then it's not and I don't get an exception. You sure that's correct behavior?

Reply:
Also the whole try block is too long, the exception is only about the timestamp, so why is everything there? A lot can move above I think, you just add or not add the metadata based on the `_timestamp_enabled` flag.

Reply:
OK, now the TimestampError will be re-raised in the grab, and everithing else is moved out from the try-block.